File: packages/py/polywrap-wasm/old-src/buffer.py
from typing import List


def write_string(
    str: str,
    dst: bytearray,
    dst_offset: int
) -> List[int]:
    str_buffer = str.encode("ascii")
    view = list(dst)
    return mem_cpy(
        str_buffer,
        0,
        view,
        dst_offset,
        len(str_buffer)
    )


def write_bytes(
    bytes: bytearray,
    dst: bytearray,
    dst_offset: int
) -> List[int]:
    bytes_view = list(bytes)
    dst_view = list(dst)
    return mem_cpy(bytes_view, 0, dst_view, dst_offset, len(bytes_view))


def read_bytes(
    from_src: bytearray,
    offset: int,
    length: int
) -> bytearray:
    buffer = bytearray()
    write_bytes(from_src[offset, offset + length], buffer, 0)
    return buffer


def read_string(
    from_src: bytearray,
    offset: int,
    length: int
):
    buffer = read_bytes(from_src, offset, length)
    return buffer.decode()


def mem_cpy(
    src: List[int],
    src_offset: int,
    dst: List[int],
    dst_offset: int,
    length: int
) -> List[int]:
    dst[0 + dst_offset: length] = src[0 + src_offset: length]
    return dst

File: packages/js/wasm/src/helpers/buffer.ts
/* eslint-disable @typescript-eslint/ban-ts-comment */
export function isBuffer(maybeBuf: unknown): maybeBuf is BufferSource {
  if (maybeBuf instanceof ArrayBuffer || ArrayBuffer.isView(maybeBuf)) {
    return true;
  } else {
    return false;
  }
}

export function writeString(
  str: string,
  dst: ArrayBuffer,
  dstOffset: number
): Uint8Array {
  const encoder = new TextEncoder();
  const strBuffer = encoder.encode(str);
  const view = new Uint8Array(dst);
  return memcpy(strBuffer, 0, view, dstOffset, strBuffer.byteLength);
}

export function writeBytes(
  bytes: ArrayBuffer,
  dst: ArrayBuffer,
  dstOffset: number
): Uint8Array {
  const bytesView = new Uint8Array(bytes);
  const dstView = new Uint8Array(dst);
  return memcpy(bytesView, 0, dstView, dstOffset, bytesView.byteLength);
}

export function readBytes(
  from: ArrayBuffer,
  offset: number,
  length: number
): ArrayBuffer {
  const buffer = new ArrayBuffer(length);
  writeBytes(from.slice(offset, offset + length), buffer, 0);
  return buffer;
}

export function readString(
  from: ArrayBuffer,
  offset: number,
  length: number
): string {
  const buffer = readBytes(from, offset, length);
  const decoder = new TextDecoder();
  return decoder.decode(buffer);
}

function memcpy(
  src: Uint8Array,
  srcOffset: number,
  dst: Uint8Array,
  dstOffset: number,
  length: number
): Uint8Array {
  // @ts-ignore
  src = (src.subarray || src.slice ? src : src.buffer) as Uint8Array;
  // @ts-ignore
  dst = (dst.subarray || dst.slice ? dst : dst.buffer) as Uint8Array;

  src = srcOffset
    ? src.subarray
      ? src.subarray(srcOffset, length && srcOffset + length)
      : src.slice(srcOffset, length && srcOffset + length)
    : src;

  if (dst.set) {
    dst.set(src, dstOffset);
  } else {
    for (let i = 0; i < src.length; i++) {
      dst[i + dstOffset] = src[i];
    }
  }

  return dst;
}
